README.md
# death

A program that predicts your death date (Learning project)

![Demonstration of CLI death](death.png)

## Installation

Download an executable from
[Releases page](https://codeberg.org/wadrodrog/death/releases)

If you're on Linux, you need to allow execution:

```shell
$ chmod +x death
```

## Usage

### As program


```
death [OPTIONS]
```

### Options

```
  -n, --name <NAME>           Your name
  -b, --birthday <BIRTHDAY>   Your birthday
  -d, --death-reasons <FILE>  Custom death reasons file
  -h, --help                  Print help
  -V, --version               Print version
```

### As library

Add dependency to `Cargo.toml`:

```toml
[dependencies]
death = { git = "https://codeberg.org/wadrodrog/death.git" }  # latest
death = { git = "https://codeberg.org/wadrodrog/death.git", tag = "0.2.0" }  # from tag
```

`src/main.rs`:

```rust
use death::{date::Date, user::User};

fn main() {
    let user = User::new(1234567890, 45, vec![String::from("lego")]);
    println!("{}", Date::today());
    println!("{} - {}", user.get_death_date(false), user.get_death_reason());
}

```

Generate docs:
```shell
$ cargo doc --open
```

## Build

You need [Cargo](https://doc.rust-lang.org/cargo/index.html) to build this
program.

```shell
$ git clone https://codeberg.org/wadrodrog/death.git
$ cd death
$ cargo build
$ ./target/debug/death
```


CHANGELOG.md
# Changelog

## [Unreleased]

### New death date formula

- Death date is now calculated using an exponential function instead of a
linear function. Smaller values will return more often than larger values.
- Pass `--linear true` option to use old linear formula.

## [0.2.0] - 2023-11-06

### Command-line arguments

```
Usage: death [OPTIONS]

Options:
  -n, --name <NAME>           Your name
  -b, --birthday <BIRTHDAY>   Your birthday
  -d, --death-reasons <FILE>  Custom death reasons file
  -h, --help                  Print help
  -V, --version               Print version
```

- Load your own death reasons with `--death-reasons <FILE>` argument. Each
death reason is on separate line.
- Added some colors to output of errors and enhanced general experience.

### API changes

This version introduces breaking API changes. You need to update your program
if you used `death` as a library. Also predictions will be different compared
to the previous version.


src/tests.rs
#[cfg(test)]
mod date;


src/tests/date.rs
#[test]
fn build_date() {
    use crate::date::{Date, ParseError};
    assert_eq!(Date::build(2015, 0, 0), Err(ParseError::InvalidMonth));
    assert_eq!(Date::build(2015, 13, 1), Err(ParseError::InvalidMonth));
    assert_eq!(Date::build(2015, 1, 32), Err(ParseError::InvalidDay));
    assert_eq!(Date::build(2015, 2, 29), Err(ParseError::InvalidDay));
    assert_eq!(Date::build(2016, 2, 30), Err(ParseError::InvalidDay));
}

#[test]
fn parse_date() {
    use crate::date::{Date, ParseError};
    // Success
    assert_eq!(
        Date::parse(&String::from("23.10.2015")), Date::build(2015, 10, 23)
    );
    assert_eq!(
        Date::parse(&String::from("23/10/2015")), Date::build(2015, 10, 23)
    );
    assert_eq!(
        Date::parse(&String::from("23-10-2015")), Date::build(2015, 10, 23)
    );
    assert_eq!(
        Date::parse(&String::from("23 10 2015")), Date::build(2015, 10, 23)
    );

    // Fail
    assert_eq!(
        Date::parse(&String::from("23\\09\\2015")),
        Err(ParseError::SeparatorNotFound)
    );
    assert_eq!(
        Date::parse(&String::from("23_09_2015")),
        Err(ParseError::SeparatorNotFound)
    );
    assert_eq!(
        Date::parse(&String::from("23092015")),
        Err(ParseError::SeparatorNotFound)
    );
    assert_eq!(
        Date::parse(&String::from("23.09.20.15")),
        Err(ParseError::InvalidPartsCount)
    );
    assert_eq!(
        Date::parse(&String::from("qwerty/asdfg.zxcvb")),
        Err(ParseError::NumberConversionError)
    );
    assert_eq!(
        Date::parse(&String::from("23.092015")),
        Err(ParseError::NumberConversionError)
    );
    assert_eq!(
        Date::parse(&String::from("20/10/-2015")),
        Err(ParseError::NumberConversionError)
    );
    assert_eq!(
        Date::parse(&String::from("32/10/2015")),
        Err(ParseError::InvalidDay)
    );
    assert_eq!(
        Date::parse(&String::from("20/13/2015")),
        Err(ParseError::InvalidMonth)
    );
    assert_eq!(
        Date::parse(&String::from("29/2/2015")),
        Err(ParseError::InvalidDay)
    );
}

#[test]
fn max_day() {
    use crate::date::Date;
    assert_eq!(Date::build(2015, 1, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 2, 1).unwrap().get_max_day(), 28);
    assert_eq!(Date::build(2016, 2, 1).unwrap().get_max_day(), 29);
    assert_eq!(Date::build(2015, 3, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 4, 1).unwrap().get_max_day(), 30);
    assert_eq!(Date::build(2015, 5, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 6, 1).unwrap().get_max_day(), 30);
    assert_eq!(Date::build(2015, 7, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 8, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 9, 1).unwrap().get_max_day(), 30);
    assert_eq!(Date::build(2015, 10, 1).unwrap().get_max_day(), 31);
    assert_eq!(Date::build(2015, 11, 1).unwrap().get_max_day(), 30);
    assert_eq!(Date::build(2015, 12, 1).unwrap().get_max_day(), 31);
}


src/cli.rs
use std::path::PathBuf;
use std::{fmt, process};
use std::io::{self, Write};

use crate::date::{self, Date, ParseError};

use clap::Parser;
use colored::*;

/// A program that predicts your death date
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    /// Your name
    #[arg(short, long)]
    pub name: Option<String>,

    /// Your birthday
    #[arg(short, long)]
    pub birthday: Option<String>,

    /// Custom death reasons file
    #[arg(short, long, value_name = "FILE")]
    pub death_reasons: Option<PathBuf>,

    /// Use linear formula to calculate the date
    #[arg(short, long)]
    pub linear: Option<bool>,
}

/// Parse command-line arguments.
pub fn parse() -> Cli {
    Cli::parse()
}

/// Print error to stderr.
///
/// If `code` is not `0`, program will close with this code.
pub fn print_error<T: fmt::Display>(error: T, code: i32) {
    eprintln!("{} {}", "error:".red(), error);
    if code != 0 {
        process::exit(code);
    }
}

/// Read string from console input.
pub fn input() -> String {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse().unwrap()
}

/// Ask user something in the same line as input.
pub fn prompt(msg: &str) -> String {
    print!("{}: ", msg);
    let _ = io::stdout().flush();
    input()
}

/// Ask user's name.
pub fn ask_name() -> String {
    prompt("Your name")
}

/// Parse birthday from string.
///
/// # Errors
///
/// Returns a string containing the reason why parsing was failed.
pub fn parse_birthday(string: &String) -> Result<Date, String> {
    let today = Date::today();
    let birthday = match Date::parse(string) {
        Ok(d) => d,
        Err(e) => {
            let msg = match e {
                ParseError::SeparatorNotFound =>
                    "Use '/', or '.', or '-', or whitespace \
                    as separator between day, month and year.",
                ParseError::InvalidPartsCount =>
                    "Invalid should be DD/MM/YYYY - day, month and year.",
                ParseError::NumberConversionError => "Invalid number.",
                ParseError::InvalidYear => "Invalid year.",
                ParseError::InvalidMonth => "Invalid month.",
                ParseError::InvalidDay => "Invalid day.",
            };
            return Err(String::from(msg));
        }
    };
    if birthday.years_from(Date::today()) >= date::MAX_AGE {
        return Err(
            String::from("Your birthday cannot be so much in the past.")
        );
    }
    if today < birthday {
        return Err(String::from("Your birthday cannot be in the future."));
    }
    Ok(birthday)
}

/// Ask user's birthday
pub fn ask_birthday() -> u8 {
    let birthday;
    loop {
        let inp = prompt("Enter your birthday (DD/MM/YYYY)");
        let _ = io::stdout().flush();
        birthday = match parse_birthday(&inp) {
            Ok(bday) => bday,
            Err(e) => {
                print_error(e, 0);
                continue;
            }
        };
        break;
    }
    birthday.years_from(Date::today()) as u8
}


src/lib.rs
use std::fs;
use std::path::PathBuf;
use std::io::{Error, ErrorKind};

pub mod date;
pub mod user;
pub mod cli;

/// Returns default death reasons.
pub fn default_death_reasons() -> Vec<String> {
    vec![
        "cars".to_string(), "illness".to_string(), "height".to_string(),
        "darkness".to_string(), "fire".to_string(), "water".to_string(),
        "nature".to_string(), "construction".to_string(),
        "electricity".to_string(), "explosions".to_string(),
        "food".to_string(), "animals".to_string(), "temperature".to_string(),
        "weapons".to_string()
    ]
}

/// Returns death reasons from file. Each reason is on separate lines. Lines
/// are trimmed of leading and trailing spaces.
///
/// If [`None`] was passed, a default death reasons returned.
///
/// # Errors
///
/// Returns [`std::io::Error`] if cannot read the file.
pub fn read_death_reasons(file_path: &Option<PathBuf>)
-> Result<Vec<String>, Error> {
    let file_path = match file_path {
        Some(x) => x,
        None => return Ok(default_death_reasons()),
    };

    let contents = fs::read_to_string(file_path)?;

    let mut res = vec![];

    for line in contents.lines() {
        let line = line.trim().to_string();
        if line.len() > 0 {
            res.push(line);
        }
    }

    if res.len() == 0 {
        Err(Error::new(ErrorKind::Other, "File is empty"))
    } else {
        Ok(res)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::{Error, ErrorKind};

    #[test]
    fn read_file() {
        // Generic file with different characters, empty lines and spaces
        let reference = vec![
            "sample1".to_string(), "sample2".to_string(),
            "sample3;".to_string(), "123".to_string(), "!@#$%^&*()".to_string(),
            "``".to_string(), "Sample With Spaces".to_string(),
            "It   Should  Trim Spaces".to_string(),
        ];

        let sample = read_death_reasons(&Some(PathBuf::from("tests/read_file.txt")))
            .unwrap();

        assert_eq!(sample, reference);

        // Empty file and file with spaces
        let error_ref = Error::new(ErrorKind::Other, "File is empty");

        match read_death_reasons(&Some(PathBuf::from("tests/spaces.txt"))) {
            Ok(_) => panic!("It is not an error as expected"),
            Err(e) => {
                assert_eq!(e.kind(), error_ref.kind());
                assert_eq!(e.to_string(), error_ref.to_string());
            },
        }

        match read_death_reasons(&Some(PathBuf::from("tests/empty.txt"))) {
            Ok(_) => panic!("It is not an error as expected"),
            Err(e) => {
                assert_eq!(e.kind(), error_ref.kind());
                assert_eq!(e.to_string(), error_ref.to_string());
            }
        }
    }
}

src/user.rs
use crate::date::{Date, self};
use crate::cli;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[derive(Debug)]
pub struct User {
    id: u64,
    age: u8,
    death_reasons: Vec<String>,
}

impl User {
    /// Returns a new user.
    pub fn new(id: u64, age: u8, death_reasons: Vec<String>) -> User {
        User { id, age, death_reasons }
    }

    /// Returns a new user from command-line arguments.
    ///
    /// If some argument was empty, default value will be used.
    ///
    /// # Errors
    ///
    /// If some argument is invalid, program will close immediately.
    pub fn from(args: &cli::Cli) -> User {
        let empty = String::from("01/01/1970");
        let birthday_string = match &args.birthday {
            Some(v) => v,
            None => &empty,
        };
        let birthday = cli::parse_birthday(&birthday_string);
        let age = match birthday {
            Ok(v) => v.years_from(Date::today()) as u8,
            Err(e) => {
                cli::print_error(e, 1);
                0
            }
        };

        let death_reasons = match crate::read_death_reasons(
            &args.death_reasons
        ) {
            Ok(v) => v,
            Err(e) => {
                cli::print_error(e, 1);
                vec![]
            },
        };

        let id = User::get_id_from_string(
            &args.name.as_deref().unwrap_or("").to_string()
        );

        User { id, age, death_reasons }
    }

    /// Set an id for user.
    pub fn set_id(&mut self, id: u64) {
        self.id = id;
    }

    /// Set an age for user.
    pub fn set_age(&mut self, age: u8) {
        self.age = age;
    }
    
    /// Get an id from string's hash.
    pub fn get_id_from_string(string: &String) -> u64 {
        let mut s = DefaultHasher::new();
        string.hash(&mut s);
        s.finish()
    }

    fn get_years_left(&self, linear: bool) -> u8 {
        if linear {
            let max_age: u64 = (date::MAX_AGE - self.age as u16) as u64;
            return (self.id % max_age + 1) as u8;
        }

        // Returns smaller values more often than larger values

        // Max y
        let max_age: f64 = (date::MAX_AGE - self.age as u16) as f64;

        // Stretch the graph horizontally to make the result more accurate
        let k: f64 = 100.0;

        // base^0=1, base^x_max=max_age
        let base: f64 = max_age.powf(1.0 / (max_age * k));

        // From 0 to max_age * k - 1
        let x: f64 = (self.id % (max_age * k) as u64) as f64;

        // f(x)=a^x
        base.powf(x) as u8
    }

    /// Returns user's predicted death reason.
    pub fn get_death_reason(&self) -> &String {
        &self.death_reasons[
            (self.id % (self.death_reasons.len() as u64)) as usize
        ]
    }

    /// Returns calculated death date of current user.
    pub fn get_death_date(&self, linear: bool) -> Date {
        let year = Date::today().year() + self.get_years_left(linear) as u16;
        let month = (self.id % 12 + 1) as u8;
        let day = 1;
        let date0 = Date::build(year, month, day).unwrap();
        let day = (self.id % date0.get_max_day() as u64 + 1) as u8;
        let date = Date::build(year, month, day).unwrap();

        date
    }
}


src/date.rs
use chrono::{Local, Datelike};
use std::{cmp, fmt};

#[derive(Debug)]
#[derive(PartialEq)]
pub enum ParseError {
    SeparatorNotFound,
    InvalidPartsCount,
    NumberConversionError,
    InvalidYear,
    InvalidMonth,
    InvalidDay,
}

#[derive(Debug)]
#[derive(PartialEq)]
#[derive(PartialOrd)]
#[derive(Eq)]
#[derive(Ord)]
#[derive(Clone, Copy)]
pub struct Date {
    year: u16,
    month: u8,
    day: u8,
}

pub const MAX_AGE: u16 = 100;

impl Date {
    /// Creates a new [`Date`] object from today's date.
    pub fn today() -> Date {
        let dt = Local::now().date_naive();
        Date {
            year: dt.year() as u16,
            month: dt.month() as u8,
            day: dt.day() as u8,
        }
    }

    /// Creates a new [`Date`] object from year, month and day.
    ///
    /// # Errors
    ///
    /// Returns [`crate::date::ParseError`] if values are invalid
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// let date = Date::build(2023, 10, 27).unwrap();
    ///
    /// assert_eq!(date.year(), 2023);
    /// assert_eq!(date.month(), 10);
    /// assert_eq!(date.day(), 27);
    /// ```
    pub fn build(year: u16, month: u8, day: u8) -> Result<Date, ParseError> {
        if year == 0 {
            return Err(ParseError::InvalidYear);
        }

        if month < 1 || month > 12 {
            return Err(ParseError::InvalidMonth);
        }
        
        if day < 1 || day > Date::max_day_of(year, month) {
            return Err(ParseError::InvalidDay);
        }

        Ok(Date { year, month, day })
    }

    /// Creates a new [`Date`] object from string.
    ///
    /// # Errors
    ///
    /// Returns [`crate::date::ParseError`] if string contains invalid date.
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// let s = String::from("27/10/2023");
    /// let date = Date::parse(&s);
    ///
    /// assert_eq!(Date::build(2023, 10, 27), date);
    /// ```
    pub fn parse(s: &String) -> Result<Date, ParseError> {
        // Find a separator
        let separators = vec!['.', '/', '-', ' '];
        let mut sep: Option<char> = None;
        for &separator in separators.iter() {
            if s.contains(separator) {
                sep = Some(separator);
                break;
            }
        }
        
        // Split into parts
        let parts;

        if let Some(value) = sep {
            parts = s.split(value);
        } else {
            return Err(ParseError::SeparatorNotFound);
        }

        // Iterate parts
        let mut numbers = vec![];
        
        for part in parts {
            let n: u16 = match part.parse() {
                Ok(n) => n,
                Err(_) => return Err(ParseError::NumberConversionError),
            };
            numbers.push(n);
        }

        if numbers.len() != 3 {
            return Err(ParseError::InvalidPartsCount);
        }

        Date::build(numbers[2], numbers[1] as u8, numbers[0] as u8)
    }

    /// Returns `true` if the year is leap.
    ///
    /// # Example
    /// ```
    /// use death::date::Date;
    ///
    /// assert!(Date::is_leap_year(2016));
    /// assert!(!Date::is_leap_year(2015));
    /// assert!(Date::is_leap_year(2000));
    /// assert!(!Date::is_leap_year(1900));
    /// ```
    pub fn is_leap_year(year: u16) -> bool {
        year % 4 == 0 && year % 100 != 0 || year % 400 == 0
    }

    /// Returns `true` if the year of current date is leap.
    /// (see [`Date::is_leap_year`])
    pub fn leap_year(&self) -> bool {
        Date::is_leap_year(self.year)
    }

    /// Returns max day of month in the year.
    ///
    /// # Example
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(Date::max_day_of(2015, 3), 31);
    /// assert_eq!(Date::max_day_of(2015, 4), 30);
    /// assert_eq!(Date::max_day_of(2015, 2), 28);
    /// assert_eq!(Date::max_day_of(2016, 2), 29);
    /// ```
    pub fn max_day_of(year: u16, month: u8) -> u8 {
        let a = vec![1, 3, 5, 7, 8, 10, 12];
        
        if month == 2 {
            if Date::is_leap_year(year) {
                return 29;
            }
            return 28;
        }

        if a.contains(&month) {
            return 31;
        }
        30
    }

    /// Returns max day of month of current date.
    /// (see [`Date::get_max_day`])
    pub fn get_max_day(&self) -> u8 {
        Date::max_day_of(self.year, self.month)
    }

    /// Returns month name
    ///
    /// # Example
    /// ```
    /// use death::date::Date;
    ///
    /// let date = Date::build(2013, 12, 11).unwrap();
    ///
    /// assert_eq!(date.get_month_name(), String::from("December"));
    /// ```
    pub fn get_month_name(&self) -> &str {
        let months = vec![
            "January", "February", "March", "April", "May", "June", "July",
            "August", "September", "October", "November", "December"
        ];
        months[(self.month - 1) as usize]
    }

    /// Returns copy of [`Date`] object with month number increased.
    ///
    /// If day was greater than next month's max day, it will be set to max day.
    ///
    /// If month was `12`, year will be increased and month set to `1`.
    ///
    /// # Example
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(
    ///     Date::build(2015, 3, 7).unwrap().next_month(),
    ///     Date::build(2015, 4, 7).unwrap()
    /// );
    /// assert_eq!(
    ///     Date::build(2015, 1, 31).unwrap().next_month(),
    ///     Date::build(2015, 2, 28).unwrap()
    /// );
    /// assert_eq!(
    ///     Date::build(2015, 12, 7).unwrap().next_month(),
    ///     Date::build(2016, 1, 7).unwrap()
    /// );
    /// ```
    pub fn next_month(&self) -> Date {
        let mut date = Date {
            year: self.year(), month: self.month(), day: self.day()
        };

        if date.month() == 12 {
            date.year += 1;
            date.month = 1;
        } else {
            date.month += 1;
        }

        date.day = date.day.clamp(1, date.get_max_day());

        date
    }

    /// Returns copy of [`Date`] object with day number increased.
    ///
    /// If day was last in current month, month will be increased and day
    /// set to `1`. If month was `12`, year will be increased, month and day
    /// set to `1`.
    ///
    /// # Example
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(
    ///     Date::build(2015, 3, 7).unwrap().next_day(),
    ///     Date::build(2015, 3, 8).unwrap()
    /// );
    /// assert_eq!(
    ///     Date::build(2015, 2, 28).unwrap().next_day(),
    ///     Date::build(2015, 3, 1).unwrap()
    /// );
    /// assert_eq!(
    ///     Date::build(2016, 2, 28).unwrap().next_day(),
    ///     Date::build(2016, 2, 29).unwrap()
    /// );
    /// assert_eq!(
    ///     Date::build(2015, 12, 31).unwrap().next_day(),
    ///     Date::build(2016, 1, 1).unwrap()
    /// );
    /// ```
    pub fn next_day(&self) -> Date {
        let mut date = Date {
            year: self.year(), month: self.month(), day: self.day()
        };

        if date.day() == date.get_max_day() {
            if date.month() == 12 {
                date.year += 1;
                date.month = 1;
            } else {
                date.month += 1;
            }
            date.day = 1;
        } else {
            date.day += 1;
        }

        date
    }

    /// Returns a number of full years from the other date.
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// let a = Date::build(1998, 5, 12).unwrap();
    /// let b = Date::build(2015, 4, 2).unwrap();
    /// let c = Date::build(2015, 5, 13).unwrap();
    ///
    /// assert_eq!(a.years_from(b), 16);
    /// assert_eq!(a.years_from(c), 17);
    /// ```
    pub fn years_from(&self, other: Date) -> u16 {
        let left = cmp::min(self.clone(), other.clone());
        let right = cmp::max(self.clone(), other.clone());
        let mut diff = right.year() - left.year();

        if right.month() < left.month() ||
        right.month() == left.month() && right.day() < left.day() {
            diff -= 1;
        }

        diff
    }

    /// Returns year value.
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(Date::build(2015, 1, 7).unwrap().year(), 2015);
    /// ```
    pub fn year(&self) -> u16 {
        self.year
    }

    /// Returns month value from 1 to 12.
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(Date::build(2015, 1, 7).unwrap().month(), 1);
    /// ```
    pub fn month(&self) -> u8 {
        self.month
    }

    /// Returns day value.
    ///
    /// # Example
    ///
    /// ```
    /// use death::date::Date;
    ///
    /// assert_eq!(Date::build(2015, 1, 7).unwrap().day(), 7);
    /// ```
    pub fn day(&self) -> u8 {
        self.day
    }
}

impl fmt::Display for Date {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} {} {}", self.day(), self.get_month_name(), self.year())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_date() {
        assert_eq!(Date::build(0, 1, 1), Err(ParseError::InvalidYear));
        assert_eq!(Date::build(2015, 0, 0), Err(ParseError::InvalidMonth));
        assert_eq!(Date::build(2015, 13, 1), Err(ParseError::InvalidMonth));
        assert_eq!(Date::build(2015, 1, 32), Err(ParseError::InvalidDay));
        assert_eq!(Date::build(2015, 1, 0), Err(ParseError::InvalidDay));
        assert_eq!(Date::build(2015, 2, 29), Err(ParseError::InvalidDay));
        assert_eq!(Date::build(2016, 2, 30), Err(ParseError::InvalidDay));
    }

    #[test]
    fn parse_date() {
        // Success
        assert_eq!(
            Date::parse(&String::from("23.10.2015")), Date::build(2015, 10, 23)
        );
        assert_eq!(
            Date::parse(&String::from("23/10/2015")), Date::build(2015, 10, 23)
        );
        assert_eq!(
            Date::parse(&String::from("23-10-2015")), Date::build(2015, 10, 23)
        );
        assert_eq!(
            Date::parse(&String::from("23 10 2015")), Date::build(2015, 10, 23)
        );

        // Fail
        assert_eq!(
            Date::parse(&String::from("23\\09\\2015")),
            Err(ParseError::SeparatorNotFound)
        );
        assert_eq!(
            Date::parse(&String::from("23_09_2015")),
            Err(ParseError::SeparatorNotFound)
        );
        assert_eq!(
            Date::parse(&String::from("23092015")),
            Err(ParseError::SeparatorNotFound)
        );
        assert_eq!(
            Date::parse(&String::from("23.09.20.15")),
            Err(ParseError::InvalidPartsCount)
        );
        assert_eq!(
            Date::parse(&String::from("qwerty/asdfg.zxcvb")),
            Err(ParseError::NumberConversionError)
        );
        assert_eq!(
            Date::parse(&String::from("23.092015")),
            Err(ParseError::NumberConversionError)
        );
        assert_eq!(
            Date::parse(&String::from("20/10/-2015")),
            Err(ParseError::NumberConversionError)
        );
        assert_eq!(
            Date::parse(&String::from("32/10/2015")),
            Err(ParseError::InvalidDay)
        );
        assert_eq!(
            Date::parse(&String::from("20/13/2015")),
            Err(ParseError::InvalidMonth)
        );
        assert_eq!(
            Date::parse(&String::from("29/2/2015")),
            Err(ParseError::InvalidDay)
        );
        assert_eq!(
            Date::parse(&String::from("1/2/0")),
            Err(ParseError::InvalidYear)
        );
    }

    #[test]
    fn max_day() {
        assert_eq!(Date::build(2015, 1, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 2, 1).unwrap().get_max_day(), 28);
        assert_eq!(Date::build(2016, 2, 1).unwrap().get_max_day(), 29);
        assert_eq!(Date::build(2015, 3, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 4, 1).unwrap().get_max_day(), 30);
        assert_eq!(Date::build(2015, 5, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 6, 1).unwrap().get_max_day(), 30);
        assert_eq!(Date::build(2015, 7, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 8, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 9, 1).unwrap().get_max_day(), 30);
        assert_eq!(Date::build(2015, 10, 1).unwrap().get_max_day(), 31);
        assert_eq!(Date::build(2015, 11, 1).unwrap().get_max_day(), 30);
        assert_eq!(Date::build(2015, 12, 1).unwrap().get_max_day(), 31);

        assert_eq!(Date::max_day_of(2015, 1), 31);
        assert_eq!(Date::max_day_of(2015, 2), 28);
        assert_eq!(Date::max_day_of(2016, 2), 29);
        assert_eq!(Date::max_day_of(2015, 3), 31);
        assert_eq!(Date::max_day_of(2015, 4), 30);
        assert_eq!(Date::max_day_of(2015, 5), 31);
        assert_eq!(Date::max_day_of(2015, 6), 30);
        assert_eq!(Date::max_day_of(2015, 7), 31);
        assert_eq!(Date::max_day_of(2015, 8), 31);
        assert_eq!(Date::max_day_of(2015, 9), 30);
        assert_eq!(Date::max_day_of(2015, 10), 31);
        assert_eq!(Date::max_day_of(2015, 11), 30);
        assert_eq!(Date::max_day_of(2015, 12), 31);
    }
}
